R/utilities-mass-balance.R

```diff
@@ -8,277 +8,238 @@
+    applicationResults <- getApplicationResults(applications)
 
-  # 2) Regroup the lumen sub-compartment but Feces
-  lumenGroupingCondition <- pathsArray[, "parentCompartmentName"] %in% "Lumen" & !pathsArray[, "subCompartmentName"] %in% "Feces"
-  pathsArray[lumenGroupingCondition, "subCompartmentName"] <- "Lumen"
-
-  # Factors could not be used for steps 1 and 2 but are needed in the next step
-  pathsArray <- as.data.frame(lapply(pathsArray, function(pathsArrayColumn) {
-    as.factor(pathsArrayColumn)
-  }))
-
-  # Aggregate Data by compound and compartment
-  simulationResultsOutputByGroup <- NULL
-
-  for (compoundLevel in levels(pathsArray[, "compoundName"])) {
-    for (compartmentLevel in levels(pathsArray[, "subCompartmentName"])) {
-      pathFilter <- pathsArray[, "compoundName"] %in% compoundLevel & pathsArray[, "subCompartmentName"] %in% compartmentLevel
-      # as.character() ensures that levels filtered out won't still be added to the aggregation
-      pathNamesFiltered <- as.character(pathsArray[pathFilter, "path"])
+    # Format to appropriate units
+    massBalanceData$Time <- ospsuite::toUnit(
+      "Time",
+      massBalanceData$Time,
+      structureSet$simulationSet$timeUnit
+    )
+    timeTicks <- getTimeTicksFromUnit(
+      structureSet$simulationSet$timeUnit,
+      timeValues = massBalanceData$Time
+    )
 
-      # cbind(..., 0) prevents rowSums to crash if only one column is filtered
-      aggregatedData <- rowSums(cbind.data.frame(
-        simulationResultsOutput$data[, pathNamesFiltered, drop = FALSE],
-        data.frame(dummyVariable = 0)
-      ))
+    metaData <- list(
+      "Time" = list(
+        dimension = "Time",
+        unit = structureSet$simulationSet$timeUnit
+      ),
+      "Amount" = list(
+        dimension = applications[[1]]$drugMass$dimension,
+        unit = applications[[1]]$drugMass$unit
+      ),
+      "NormalizedAmount" = list(
+        dimension = applications[[1]]$drugMass$dimension,
+        unit = "fraction of drugmass"
+      )
+    )
 
-      if (max(aggregatedData) > 0) {
-        simulationResultsOutputByGroup <- rbind.data.frame(
-          simulationResultsOutputByGroup,
-          cbind.data.frame(
-            "Time" = toUnit("Time", simulationResultsOutput$data[, "Time"], structureSet$simulationSet$timeUnit),
-            "Amount" = aggregatedData,
-            "NormalizedAmount" = aggregatedData / drugMass,
-            "Legend" = paste0(compoundLevel, " - ", compartmentLevel)
-          )
-        )
-      }
-    }
-  }
+    # The function approx aims at extrapolating
+    # the total drug mass data on the time points of the mass balance data
+    massBalanceData$DrugMass <- approx(
+      x = ospsuite::toUnit(
+        "Time",
+        applicationResults$time,
+        structureSet$simulationSet$timeUnit
+      ),
+      y = applicationResults$totalDrugMass,
+      xout = massBalanceData$Time,
+      method = "constant",
+      rule = 2
```

Comment:
can you extend the comment explaining what "rule = 2" means?

Reply:
Yes, I'll include that in the code as well.
To keep track, I'll also write here: the argument `rule` defines what happens in case you extrapolate `yout` outside of the smallest/biggest `x` value. `rule=1` will predict `NA` and `rule=2` will predict `y` value of smallest/biggest `x` value.